### ISSUE TITLE
Doubts regarding the paper/implementation

### ISSUE THREAD

**Issue by divyagupta25** [fywalter/simptc#2]
Hi, 

I have some doubts regarding the paper/implementation. Could you please help to resolve them? 

1. Why is the test set included for the GMM training? Should the test set not be held-out? 
2. Which model is used for Vanilla prompting?
3. In T5-3B, is the average of all token embeddings from the last hidden state used as the sentence embedding? 
https://github.com/fywalter/simptc/blob/eb337d5fb4f4da561cd5224788c439ae81c7376b/encode_dataset.py#L40
          `model_output = model.encoder(**encoded_input, return_dict=True)`
          `sent_embedding = (model_output.last_hidden_state * encoded_input['attention_mask'].unsqueeze(-1)).sum(dim=-2) / encoded_input['attention_mask'].sum(dim=-1)`
          `sent_embedding = sent_embedding.cpu().detach().numpy()`
          
   Is this to be used in the else block? 
4. Regarding the embedding space of T5-3B, I plotted for IMDb train set using the shared code. 
![imdb PCA_conceptnet_1000_roberta_large_2](https://github.com/fywalter/simptc/assets/30105023/f7f7186b-82f2-4096-b0cd-056e68ed3550)
![imdb PCA_conceptnet_1000_t5-3b_2](https://github.com/fywalter/simptc/assets/30105023/01a25302-9306-4ec2-ba54-eed509ecf3d2)
The first is for SimCSE-RL and second is for T5-3B. Although there is no clear "winner" in this case, but I wanted to ask if you observed more separability or any other trait in the T5 embedding space? 
In a custom dataset that I am using, I am observing lower separability in both 2-D and 3-D for T5-3B as compared to SimCSE-RL. 
5. In Appendix I.4, it is written that "A 10-shot setting improves SimPTC from 52.0 to 89.2 on Subj". In which step of SimPTC are those 10-shots used? Could you please elaborate on this?
6. Could you please explain what these mean?
https://github.com/fywalter/simptc/blob/eb337d5fb4f4da561cd5224788c439ae81c7376b/utils.py#L162-L164
7. This is a broader question about the approach - Would it be correct to say that it all comes down to the PLM's ability to separate in the embedding space? 
In the paper, it is said that for abstract classes, the right initialisation cannot be found from E&M but the data points are separated in the embedding space. 
What if it is the reverse? That is, say a small minority class is entangled with others in the embedding space. Then, there would be no way to improve the performance on that class no matter how good the initialisation is? 
Not sure if I am making sense here 

Thanks for your time!

**Comment by divyagupta25** [comment]
@fywalter Could you please help me with any of these questions? 
Thanks in advance!

**Comment by fywalter** [comment]
Hi,

Thanks for your questions.

> 1. Why is the test set included for the GMM training? Should the test set not be held-out?

One of our main arguments is that when it comes to zero-shot learning, clustering the text representations might be better than directly prompting. Specifically, in zero-shot prompting, we classify test data using some prompts. We aim to show that instead of doing this, these test data are well-clustered already in many text classification tasks and we can use the cluster assignments as the class assignments. Since the embedding spaces usually have high dimensions and the test set size are usually limited, using more unlabeled traininging data will further improve the clustering results. But you are right that the most strict way in benchmarking the text classification performance should be to only use the unlabeled training data, although as the clustering process is parameter-free we expect the results of using only the training data to have minimum difference to the one include the test data in the clustering process.

> 2. Which model is used for Vanilla prompting?

We use RoBERTa-large for Vanilla prompting, which is used to initialize cluster assignments for other LMs as well.

>3. In T5-3B, is the average of all token embeddings from the last hidden state used as the sentence embedding?

Yes. What we do is the following:

```python
def encode_sentence(text, model, tokenizer):
    encoded_input = tokenizer(text, max_length=512, truncation=True, return_tensors="pt")
    encoded_input.to(device)
    with torch.no_grad():
        output = model.encoder(
            input_ids=encoded_input["input_ids"], 
            attention_mask=encoded_input["attention_mask"], 
            return_dict=True
        )
        sent_embedding = torch.mean(output.last_hidden_state, dim=1)
    return torch.squeeze(sent_embedding).cpu().detach().numpy()
```

>4. Regarding the embedding space of T5-3B, I plotted for IMDb train set using the shared code...

Great question. There are several points to be made. First, 2d visualizations do not always give you the whole story as the clusters might separate more clearly from another angle. Second, visualizing the data can sometimes be misleading in telling you which embedding space is the "winner". As discussed in Figure 7 and Table 6 in the paper, the clusters are more separable in the sent-RL embedding space but RL gets higher score. This shows that we should not solely focusing on the "observing separability". As the LMs are not optimized to make these clusters more separable during their training, in terms of visual separability, we should expect that they are not as good as the sentence embedding models. But nevertheless, they can have good clustering results.

>5. In Appendix I.4, it is written that "A 10-shot setting improves SimPTC from 52.0 to 89.2 on Subj". In which step of SimPTC are those 10-shots used? Could you please elaborate on this?

We average the embeddings of 10-shots to initialize the cluster centers.

>6. Could you please explain what these mean?

I am glad that you read the code really carefully. These are for early experiments where we check whether, after clustering, each cluster is correctly mapped to the corresponding class. It can be that (for example, due to bad initialization, or that several classes are indistinguishable in the embedding space) after some steps, several classes are merged into one single cluster.

>7. This is a broader question about the approach - Would it be correct to say that it all comes down to the PLM's ability to separate in the embedding space?
>   In the paper, it is said that for abstract classes, the right initialisation cannot be found from E&M but the data points are separated in the embedding space.
>   What if it is the reverse? That is, say a small minority class is entangled with others in the embedding space. Then, there would be no way to improve the performance on that class no matter how good the initialisation is?
>   Not sure if I am making sense here

I think "that it all comes down to the PLM's ability to separate in the embedding space" is correct. From another angle, being able to perform zero-shot prompting means that the model can distinguish text of different classes (but not necessarily in the embedding space), and we find that for many text classification tasks, it is indeed able to separate texts of different classes.

For the second part of the question, I am not sure I correctly understand your question. I agree that both cases can happen in reality depending on what LM we are using and what task we are dealing with.

Best,
